client/src/classes/itemtemplatestore.ts

```diff
@@ -0,0 +1,37 @@
+import { Item } from './../classes/item';
+
+export class ItemTemplateStore {
+
+  private static store = {};
+
+  public static addTemplate(item: Item): void {
+    if (this.containsTemplate(item)) return;
+
+    let template = this.createTemplateFromItem(item);
+    this.store[template.getName()] = template;
+  }
+
+  public static getTemplates(): Item[] { return Object.values(this.store); }
+
+  public static getTemplate(item: Item): Item { 
+    let name = Object.keys(this.store).find(name => { return name == item.getName() });
+    return name != null ? this.store[name] : null;
+  }
+
+  public static getTemplateByName(name: string): Item {
+    return Object.keys(this.store).indexOf(name) > -1 ? this.store[name] : null;
+  }
+
+  public static updateTemplate(item: Item, perform: any): void {
+    let template: Item = this.getTemplate(item);
+    perform(template);
+  }
+
+  private static containsTemplate(item: Item): boolean {
+    return Object.keys(this.store).indexOf(item.getName()) > -1;
+  }
+
+  private static createTemplateFromItem(item: Item): Item {
+    return new Item(item.getName(), false, item.department, true);

```

Comment:
Just noticed this, that that true value is essentially your "sentinel value" that prevents an application halting infinite loop. I would try to avoid patterns that result in the need for such values.

I think a lot of the challenges you're working around here have ultimately to do with the fact that you're trying to represent two different concepts via the same time.

If you want ItemTemplates, have ItemTemplates. Let them be able to generate items. Also let them be able to look at an item and generate a template from it. Items themselves shouldn't need to know templates even exist.